README.md
# Beautiful imports

<p
    align="center"
    style="text-align: center;"
>
    <img
        src="https://raw.githubusercontent.com/abderox/BEAUTIFUL-IMPORTS/main/icons/beautifulImports.png"
        alt="Beautiful imports"
    />
</p>

## DISCLAIMER

- 👋 Please pardon me, as this is my first time creating an extension for Visual Studio Code. I have limited knowledge and expertise, so I appreciate your patience.
- 🔔 It is recommended to format your document using the widely-known shortcut `Alt+Shift+F` before utilizing this extension `ctrl+Shift+J`.
- ⚠️ Please note that the extension currently only formats inline statements. We are working on fixing this issue in upcoming updates.

## Overview

- This extension is designed to format JavaScript elements that may not be formatted by other formatters.
- Although it was initially intended to format imports, it has expanded its capabilities.

## Features

- The Beautiful imports extension is designed to improve the readability of JavaScript files by formatting imports.
- In particular, it formats inline imports with more than `3 (default value)` elements to be displayed vertically, making it easier to read and understand the imports in your code.
- Automatic sorting for imports and exports.
- Supports both import and require statements.
- Formats html tags ,React Componenets' attributes with more or equal to `3 (default value)` elements to be displayed vertically.

## Unreleased

- Formatting Selected code only to avoid problems.
- Removes unused imports.
- Formating imports on save.

## Release Notes

### 1.0.4

- Fixing some issues about (FileType not supported error).
- Components like `sx= {{...}}` can be formatted.
- Please keep in mind that the extension formats only inline statements. For example, in this case, it will only handle the first line and may result in the removal of `={{`.
  
- _It transforms the line below  :_

```typescript
    <Button variant="outlined" color="primary" component="label" sx={{
```

- _Into this which will cause you problems:_

```typescript
    <Button
              variant="outlined"
              color="primary"
              component="label"
              sx
```

- _It is recommended for now to apply it only on statements like_

```typescript
    <Button variant="outlined" color="primary" component="label" sx={{...}} />
```

- _which will result in_
  
  ```typescript
      <Button
                variant="outlined"
                color="primary"
                component="label"
                sx={{...}}
      />
  ```

### 1.0.3

- Now it can formats React Componenets' attributes and props with more or equal to `3`
_It transforms this_

```typescript
< Component  className = 'class' value1={val} value2={val2} value3=50 disabled  />
```

_To this_

```typescript
<Component
              className='class'
              value1={val}
              value2={val2}
              value3=50
              disabled
/>
```

### 1.0.2

- No changes within the extension

### 1.0.1

```diff
+ It is able now to format exports both for `commonJS` & `moduleJS`;
+ It is able now to format import (as a list) with its different known ways as follows 
```

```typescript
import {a, b, c} from 'module/path'
import {a as a1, b as b1, c as c1} from 'module/path'
import defaultExport, {a, b, c} from 'module/path'
import defaultExport, {a as a1, b as b1, c as c1} from 'module/path'
```

### 1.0.0

Initial release of Beautiful imports extension.

## Usage

<p
    align="center"
    style="text-align: center;"
>
    <img
        src="https://raw.githubusercontent.com/abderox/BEAUTIFUL-IMPORTS/main/github/extension-v1.0.3.gif"
        alt="Inline imports"
    />
</p>

- _From this_

```typescript
import {bmethod5 , cmethod6 ,amethod7, method8} from "package"

const {bmethod5 , cmethod6 ,amethod7, method8} =require ("package")

import defaultExport, { a as a1 ,b as b1, c as c1 } from "module/path"

export let Amodule = { method1: methodA, method2: methodB, method3: methodC};
```

- _To this_

```typescript
import {
  amethod5,
  bmethod6,
  cmethod7,
  method8
} from "package";

const {
  amethod5,
  bmethod6,
  cmethod7,
  method8
} = require("package");

import defaultExport, {
  a as a1,
  b as b1,
  c as c1
} from "module/path";

export let Amodule = {
  method1: methodA,
  method2: methodB,
  method3: methodC
};
```

## How to install it locally Without getting to market place💻

- Go to [Release](https://github.com/abderox/BEAUTIFUL-IMPORTS/releases)
- Download the release you want `.vsix`
- Open VS Code.
- Press `Ctrl+Shift+P` (Windows, Linux) or `Cmd+Shift+P` (macOS) to open the command palette.
- Type Extensions: `Install from VSIX` in the command palette and select it.
- Navigate to the location of the `.vsix` file on your computer and select it.
- VS Code will install the extension and prompt you to restart the editor.
- If you have any unsaved work, be sure to save it before restarting.

## Requirements

- Before installing and using the extension, you must have Node.js and npm (Node Package Manager) installed on your system. You can download the latest version of Node.js from the official website: <https://nodejs.org/en/download/>.

- Additionally, you should have Visual Studio Code version 1.75.0 or later installed. You can download the latest version from the official website: <https://code.visualstudio.com/download>.

## Extension Settings

The extension contributes a command called `extension.formatImports` that can be triggered using the keyboard shortcut `Ctrl+Shift+J` (or `Cmd+Shift+J` on Mac). It also adds a context menu item to the editor that can be used to trigger the same command. The context menu item is available when a `JavaScript` or `TypeScript` file is open in the editor.

For example:

This extension contributes the following settings:

- `beautifulImports.enable`: Enable/disable this extension.

- `beautifulImports.maxInlineImports`: The number of inline imports allowed before they are formatted vertically. Default is `3`.

- `beautifulImports.IndentsMaxNumber` : This number defines how many white spaces to put counting from the end of tagName . Deafault is `14`.

- `beautifulImports.EnableForComponents` : This setting enables/disables the extension for formatting components. Default is `false`.

## Known Issues

- ⚠️ This is  just a prototype, it will be replaced or removed sooner.
- In rare cases, formatting may not work correctly if the import statements are not formatted correctly in the first place.
- `Error: Illegal value for 'line'` may be thrown for some reason.

**Enjoy!** 😄


CHANGELOG.md
# Change Log

## Unreleased

- Formatting Selected code only to avoid problems.
- Removes unused imports.
- Formating imports on save.

## Release Notes

### [1.0.4] - 2022-02-23

- Fixing some issues about (FileType not supported error).
- Components with arrributes like `sx= {{...}}` can be formatted.
- Please keep in mind that the extension formats only inline statements. For example, in this case, it will only handle the first line and may result in the removal of `={{`.
  
- _It transforms the line below  :_

```typescript
    <Button variant="outlined" color="primary" component="label" sx={{
```

- _Into this which will cause you problems:_

```typescript
    <Button
              variant="outlined"
              color="primary"
              component="label"
              sx
```

- _It is recommended for now to apply it only on statements like_

```typescript
    <Button variant="outlined" color="primary" component="label" sx={{...}} />
```

- _which will result in_
  
  ```typescript
      <Button
                variant="outlined"
                color="primary"
                component="label"
                sx={{...}}
      />
  ```

### [1.0.3] - 2022-02-22

- Now it can formats React Componenets' attributes and props with more or equal to `3`
- It is recommended to format your document using the widely-known shortcut `Alt+Shift+F` before utilizing this extension `ctrl+Shift+J`.
- _It transforms this_

```typescript
< Component  className = 'class' value1={val} value2={val2} value3=50 disabled  />
```

- _To this_

```typescript
<Component
              className='class'
              value1={val}
              value2={val2}
              value3=50
              disabled
/>
```

### [1.0.2] - 2022-02-22

- No changes within the extension
- Changing Extension categories from other to Formatters

### [1.0.1] - 2022-02-22

```diff
+ It is able now to format exports both for `commonJS` & `moduleJS`;
+ It is able now to format import (as a list) with its different known ways as follows 
```

```typescript
import {a, b, c} from 'module/path'
import {a as a1, b as b1, c as c1} from 'module/path'
import defaultExport, {a, b, c} from 'module/path'
import defaultExport, {a as a1, b as b1, c as c1} from 'module/path'
```

## [1.0.0] - 2022-02-20

- Supports automatic sorting and formatting of import statements in JavaScript files
Features
- Supports both import and require statements

## Feautures

- The Beautiful imports extension is designed to improve the readability of JavaScript files by formatting imports.
- In particular, it formats inline imports with more than `3 (default value)` elements to be displayed vertically, making it easier to read and understand the imports in your code.
- Automatic sorting for imports and exports.
- Supports both import and require statements.
- Formats html tags ,React Componenets' attributes with more or equal to `3 (default value)` elements to be displayed vertically.

## Known Issues

- In rare cases, formatting may not work correctly if the import statements are not formatted correctly in the first place.
- `Error: Illegal value for 'line'` may be thrown.
- ⚠️ This is  just a prototype, it will be replaced or removed sooner.


src/format.js
const vscode = require('vscode');

const {
    supportedFileType,
    pushToArray,
    processLines
} = require('./utils');

const isEnabled = vscode.workspace.getConfiguration("beautifulImports").get("enabled");


function formatImports() {

    if (!isEnabled) {
        vscode.window.showErrorMessage(`beautifulImports Extension is disabled`);
        return;
    }

    const editor = vscode.window.activeTextEditor;
    const document = editor.document;

    let imports = [];
    let requires = [];
    let exports_ = [];
    let mexports = [];
    let tags = [];

    if (!editor) { 
        return;
    }

    if (document.lineCount === 0) {
        vscode.window.showWarningMessage(`This file is empty`);
        return;
    }

    for (let i = 0; i < document.lineCount; i++) {
        try {

            const line = document.lineAt(i);

            if (!line) {
                continue;
            }

            const { import_, export_, moduleExport, require_, tag } = pushToArray(line);

            if (import_ != null) {
                imports.push(import_);
            }
            if (require_ != null) {
                requires.push(require_);
            }
            if (export_ != null) {
                exports_.push(export_);
            }
            if (moduleExport != null) {
                mexports.push(moduleExport);
            }
            if (tag != null) {
                tags.push(tag);
            }


        } catch (error) {
            console.error(`Error on line ${i}: ${error}`);
            vscode.window.showErrorMessage(`Error on line ${i}:  ${error}`);

        }

    }

    let changes = processLines(imports, requires, exports_, mexports, tags);

    if (Array.isArray(changes) && changes.length != 0) {
        const edit = new vscode.WorkspaceEdit();
        edit.set(document.uri, changes);
        changes = []
        vscode.window.withProgress({
            location: vscode.ProgressLocation.Notification,
            title: 'Formating elements',
            cancellable: false
        }, (progress, token) => {
            token.onCancellationRequested(() => {
                console.log("User canceled the long running operation");
            });

            progress.report({ message: "Please wait, formatting may take a while..." });

            return vscode.workspace.applyEdit(edit);
        }).then((success) => {
            if (success) {
                vscode.window.showInformationMessage(`Elements are now Beautiful!`);
            } else {
                vscode.window.showErrorMessage(`beautifulImports Extension Failed`);
            }
        });
    }
}

function activate(context) {

    vscode.workspace.onDidChangeTextDocument((event) => {
        // Check if the event is an undo event
        if (event.contentChanges[0].rangeLength > 0 && event.contentChanges[0].text === "") {
            console.log("Undo detected. Not formatting imports.");
            return;
        }


    });
    const disposable = vscode.commands.registerCommand('extension.formatImports', formatImports);

    context.subscriptions.push(disposable);


}


function deactivate() {
    vscode.window.showErrorMessage(`beautifulImports Extension Deactivated`);
    context.subscriptions.map(e => e.dispose())

}

module.exports = {
    activate,
    deactivate
};


src/utils/constants.js

const IMPORT_STATEMENT_REGEX = /^import\s*((\w+\s*),\s*)?{\s*((\w+\s*as\s*\w+|\w+)(\s*,\s*(\w+\s*as\s*\w+|\w+))*)\s*}\s*from\s*["'](.+)["']\s*;?$/;
const REQUIRE_STATEMENT_REGEX = /^(const|let|var)?\s*{\s*((\w+)(\s*,\s*(\w+))*)\s*}\s*=\s*require\s*\(\s*['"](.+)['"]\s*\)/;
const EXPORT_STATEMENT_REGEX = /^export\s*((const|let|var)\s+(\w+)\s*=)?\s*{\s*((\w+\s*:\s*\w+|\w+)(\s*,\s*(\w+\s*:\s*\w+|\w+))*)\s*}\s*;?$/;
const MODULE_EXPORT_REGEX =/^module\.exports\s*=?\s*{\s*((\w+\s*:\s*\w+|\w+)(\s*,\s*(\w+\s*:\s*\w+|\w+))*)\s*}\s*;?$/;
const TAG_REGEX = /^(\s*)<\s*(\w+)\s*((?:\w+\s*=.*?(?=[\s/]>))*)\s*(\/?>?)/;
const TAG_ATTRIBUTES =  /((\w+)\s*=\s*("[^"]*"|'[^']*'|\{.*?\}|\d+)|\w+\s*)/g;

const SUPPORTED_FILE_TYPES = [".js",".jsx",".ts",".tsx"];


module.exports = {
    IMPORT_STATEMENT_REGEX,
    REQUIRE_STATEMENT_REGEX,
    EXPORT_STATEMENT_REGEX,
    MODULE_EXPORT_REGEX,
    TAG_REGEX,
    TAG_ATTRIBUTES,
    SUPPORTED_FILE_TYPES,
}

src/utils/index.js
const vscode = require('vscode');
const path = require('path');

const {
    IMPORT_STATEMENT_REGEX,
    REQUIRE_STATEMENT_REGEX,
    EXPORT_STATEMENT_REGEX,
    MODULE_EXPORT_REGEX,
    TAG_REGEX,
    TAG_ATTRIBUTES,
    SUPPORTED_FILE_TYPES, } = require('./constants');

function supportedFileType() {
    const filePath = vscode.workspace.textDocuments[0].uri.fsPath;
    const fileExtension = path.extname(filePath);
    return SUPPORTED_FILE_TYPES.includes(fileExtension) ? fileExtension : null;
}

function handelClosingTag(lineText) {
    const str = lineText.trimRight();
    return str.slice(-2) === "/>" ? str.slice(0, -2) + " />" : str.slice(-1) === ">" ? str.slice(0, -1) + " >" : str + " >";
}

function breakIntoElements(lineText, line) {

    const importStatement = lineText.trim().match(IMPORT_STATEMENT_REGEX);
    const requireStatement = lineText.trim().match(REQUIRE_STATEMENT_REGEX);
    const exportStatement = lineText.trim().match(EXPORT_STATEMENT_REGEX);
    const moduleExportStatement = lineText.trim().match(MODULE_EXPORT_REGEX);
    const tagStatementCheck = handelClosingTag(lineText);
    const tagStatement = tagStatementCheck.match(TAG_REGEX);
    console.log("🚀 ~ file: index.js:26 ~ breakIntoElements ~ tagStatement:", tagStatement)



    if (importStatement) {
        const [, , default_, elements] = importStatement;
        console.log("🚀 ~ file: index.js:23 ~ breakIntoElements ~ default_:", default_)
        return {
            type: 'import_',
            elements: { line, default_, elements, packageName: importStatement[importStatement.length - 1] }
        }
    } else if (requireStatement) {
        const [, key, elements] = requireStatement;
        return {
            type: 'require_',
            elements: { line, key, elements, packageName: requireStatement[requireStatement.length - 1] }
        }
    } else if (exportStatement) {
        const [, , key, name, elements] = exportStatement;
        return {
            type: 'export_',
            elements: { line, key, name, elements, packageName: exportStatement[exportStatement.length - 1] }
        }
    } else if (moduleExportStatement) {
        const [, elements] = moduleExportStatement;
        return {
            type: 'moduleExport',
            elements: { line, elements, packageName: moduleExportStatement[moduleExportStatement.length - 1] }
        }
    } else if (tagStatement) {
        const iEnabled = vscode.workspace.getConfiguration('beautifulImports').get('EnableForComponents');
        console.log("🚀 ~ file: index.js:63 ~ breakIntoElements ~ iEnabled:", iEnabled)
        

        if (!iEnabled) {

            vscode.window.showInformationMessage(
                "😇 Components have been detected, but please note that the 'EnableForComponents' feature "+
                "is currently deactivated by default. This is because it is still under development. "+
                "So, not to potentially cause issues with your code, it is recommended to take a peep at the README. "+
                "You may  activate this feature from the settings."
            );

            return {
                type: null,
                elements: null
            }
        }

        const [, sapce, tagName, attributes] = tagStatement;

        if (!attributes) {
            return {
                type: null,
                elements: null
            }
        }
        let closingTag_ = lineText.trimRight().slice(-1).includes('>') ? ">" : "";
        closingTag_ = lineText.trimRight().slice(-2).includes('/>') ? '/>' : closingTag_;
        return {
            type: 'tag',
            elements: { line, indent: " ".repeat(sapce.split(" ").length - 1), tagName, attributes, closingTag: closingTag_ }
        }
    } else {
        return {
            type: null,
            elements: null
        }
    }
}

function pushToArray(line) {


    const statements = {
        import_: null,
        require_: null,
        export_: null,
        moduleExport: null,
        tag: null
    };

    if (!line) {
        return statements;
    }

    let lineText = line.text || line.b;

    const { type, elements } = breakIntoElements(lineText, line);

    if (statements.hasOwnProperty(type)) {
        statements[type] = elements;
    }

    return statements;

}


function processTagElements(maxElementsPerLine, elements, spaces) {
    let [...attributes] = elements.match(TAG_ATTRIBUTES);
    attributes = attributes.map(e => e.trim());
    let lines = []
    // check if elements >= 3
    if (attributes.length < maxElementsPerLine) {
        return lines;
    }
    console.log("🚀 ~ file: index.js:99 ~ processTagElements ~ attributes:", attributes)
    const indetsMaxNumber = vscode.workspace.getConfiguration('beautifulImports').get('IndentsMaxNumber') || 14;
    const indent = " ".repeat(indetsMaxNumber) + spaces;

    for (let i = 0; i < attributes.length; i += maxElementsPerLine) {
        let lineElements = attributes.slice(i, i + maxElementsPerLine);
        let formatedLines = lineElements.map(e => {
            const index = e.indexOf('=');
            if (index > -1) {
                const key = e.slice(0, index);
                const value = e.slice(index + 1);
                return `${indent}${key}=${value}`;
            }
            else {
                return `${indent}${e}`;
            }

        })

        lines.push(`${formatedLines.join('\n')}`);
    }
    return lines;

}

function processElements(maxElementsPerLine, elements) {

    const elementArray = elements.split(',').map(e => e.trim());
    let lines = []
    // check if elements >= 3
    if (elementArray.length < maxElementsPerLine) {
        return lines;
    }


    for (let i = 0; i < elementArray.length; i += maxElementsPerLine) {
        let lineElements = elementArray.slice(i, i + maxElementsPerLine);
        lineElements.sort();

        let formatedLines = lineElements.map(e => {
            if (e.includes(':')) {
                const [key, value] = e.split(':').map(it => it.trim());
                return `${key}: ${value}`;
            } else if (e.includes('as')) {
                const [key, value] = e.split('as').map(it => it.trim());
                return `${key} as ${value}`;
            }
            else {
                return e;
            }

        })

        lines.push(`  ${formatedLines.join(',\n  ')}`);
    }
    return lines;
}


function processLines(imports = [], requires = [], exports = [], mexports = [], tags = []) {
    const maxElementsPerLine = parseInt(vscode.workspace.getConfiguration("beautifulImports").get("maxInlineImports"));
    let changes = []

    if (!maxElementsPerLine) {
        vscode.window.showErrorMessage(`Invalid value for maxInlineImports`);
        return;
    }

    if (imports.length > 0) {
        for (const importStatement of imports) {
            const { line, default_, elements, packageName } = importStatement;
            let lines = processElements(maxElementsPerLine, elements);
            if (lines.length === 0) { continue; }
            // Replace the import statement line with the reformatted lines
            const newText = `import ${default_ ? default_ + "," : ""} {\n${lines.join(',\n')}\n} from "${packageName.trim()}";\n`;
            changes.push(new vscode.TextEdit(line.range, newText));
        }

    }

    if (requires.length > 0) {
        for (const requireStatement of requires) {
            const { line, key, elements, packageName } = requireStatement;
            let lines = processElements(maxElementsPerLine, elements);
            if (lines.length === 0) { continue; }
            // Replace the require statement line with the reformatted lines
            const newText = `${key} {\n${lines.join(',\n')}\n} = require("${packageName.trim()}");\n`;
            changes.push(new vscode.TextEdit(line.range, newText));
        }
    }

    if (exports.length > 0) {
        for (const exportStatement of exports) {
            const { line, key = null, name = null, elements } = exportStatement;
            let lines = processElements(maxElementsPerLine, elements);
            if (lines.length === 0) { continue; }
            // Replace the export statement line with the reformatted lines
            const newText = `export ${key || ""} ${name ? name + " =" : ""} {\n${lines.join(',\n')}\n};\n`;
            changes.push(new vscode.TextEdit(line.range, newText));
        }

    }

    if (mexports.length > 0) {
        for (const mexportStatement of mexports) {
            const { line, elements } = mexportStatement;
            let lines = processElements(maxElementsPerLine, elements);
            if (lines.length === 0) { continue; }
            // Replace the module export statement line with the reformatted lines
            const newText = `module.exports = {\n${lines.join(',\n')}\n};\n`;
            changes.push(new vscode.TextEdit(line.range, newText));
        }
    }

    if (tags.length > 0) {
        for (const tagStatement of tags) {
            const { line, indent, tagName, attributes, closingTag } = tagStatement;
            let lines = processTagElements(maxElementsPerLine, attributes, indent);
            if (lines.length === 0) { continue; }
            // Replace the tag statement line with the reformatted lines
            const newText = `${indent}<${tagName.trim()}\n${lines.join('\n')}\n${closingTag ? indent + closingTag.trim() : ""}`;
            changes.push(new vscode.TextEdit(line.range, newText));
        }
    }



    return changes;
}


module.exports = {
    supportedFileType,
    pushToArray,
    processLines
}